src/unistd.rs

```diff
@@ -436,37 +450,65 @@ pub fn dup(oldfd: RawFd) -> Result<RawFd> {
 /// specified fd instead of allocating a new one.  See the man pages for more
 /// detail on the exact behavior of this function.
 #[inline]
-pub fn dup2(oldfd: RawFd, newfd: RawFd) -> Result<RawFd> {
-    let res = unsafe { libc::dup2(oldfd, newfd) };
+pub fn dup2<Fd: AsFd>(oldfd: Fd, newfd: &mut OwnedFd) -> Result<()> {
+    let res =
+        unsafe { libc::dup2(oldfd.as_fd().as_raw_fd(), newfd.as_raw_fd()) };
+
+    Errno::result(res).map(drop)
+}
+
+#[inline]
+pub fn dup2_raw<Fd1: AsFd, Fd2: AsRawFd + IntoRawFd>(
```

Comment:
If this operates on a `RawFd` it should be `unsafe`. Otherwise it could be passed a fd corresponding to a file open elsewhere, and break soundness assumptions in the code using  it.

Reply:
No, that's why it requires `IntoRawFd`. If the call fails, you had to pass in your FD by value which means it gets dropped inside `dup2_raw`. If it succeeds, we give you back the same fd. Of course if you don't want the safety stuff, then you pass in a `RawFd` and `IntoRawFd` will do nothing.

Reply:
In the current Rust version `RawFd` implements `AsRawFd` and `IntoRawFd`, so this API can easily be used with an arbitrary integer. So you could for instance call `close(dup2_raw(some_fd, 0).unwrap())`. Allowing safe code to close stdin or any other fd. This shouldn't be possible in safe code without an `OwnedFd`.

This function is only safe if the caller either owns fd2, or knows the fd isn't used.

So I think the function should be an `unsafe fn`, with a doc comment describing when it is safe.

Reply:
Maybe? When to require unsafe is unclear IMO. BTW, your example can just be `close(0)`—no need for dup. I think our goal should be to not break BorrowedFd or OwnedFd. If you want to do weird stuff with RawFd, then I feel like that's up to you.

Reply:
https://rust-lang.github.io/rfcs/3128-io-safety.html

> Functions accepting arbitrary raw I/O handle values ([RawFd](https://doc.rust-lang.org/stable/std/os/unix/io/type.RawFd.html), [RawHandle](https://doc.rust-lang.org/stable/std/os/windows/io/type.RawHandle.html), or [RawSocket](https://doc.rust-lang.org/stable/std/os/windows/io/type.RawSocket.html)) should be unsafe if they can lead to any I/O being performed on those handles through safe APIs.

Basically any function that can operate on a `RawFd` should be unsafe. So to conform to IO safety, `close` should either be unsafe, or it should take a `OwnedFd` and consume it.

This is definitely awkward for a crate like `nix` though, since conforming to this necessitates breaking changes to basically all of the API. Perhaps these functions in `nix` should have been unsafe already, but ideally Rust would have had clear rules about IO safety before 1.0.

Reply:
> This is definitely awkward for a crate like nix though

Yeah, I think that's the crux of the matter. Nix is trying to be more low-level than the stdlib, so requiring unsafe is probably going to be annoying. I also think you can argue that we're following the definition. It says "can lead to" which to me means that if you store the RawFd somewhere and then use it later (like in a BorrowedFd or File), the function that does the storing should be marked unsafe, but if you immediately use the RawFd then you're good. Otherwise anytime a RawFd is even mentioned you'd need unsafe which doesn't make sense to me.

All of that said, I wouldn't be opposed to marking `close` as unsafe. Interestingly enough, dup is perfectly safe according to "Rust code has I/O safety if it's not possible for that code to cause invalid I/O operations" because anything you throw at dup will work (if the fd doesn't exist, it'll be allocated, if it does exist, it'll be closed).

Reply:
Other than the breaking API changes, and interop with other things using `RawFd`, it seems like it should be fairly ergonomic to work with a safe API that uses `OwnedFd`/`BorrowedFd`/`AsFd` and mostly doesn't expose `RawFd` anywhere in the API (`dup2_raw` is a special case here).

> Interestingly enough, dup is perfectly safe according to "Rust code has I/O safety if it's not possible for that code to cause invalid I/O operations"Interestingly enough, dup is perfectly safe according to "Rust code has I/O safety if it's not possible for that code to cause invalid I/O operations"

This is subtle, but a safe `dup` taking a `RawFd` and returning an `OwnedFd` could be used to trigger memory unsafety with the sort of things envisioned in the IO safety RFC. In particular:

> And in specialized circumstances, violating I/O safety could even lead to violating memory safety. For example, in theory it should be possible to make a safe wrapper around an mmap of a file descriptor created by Linux's [memfd_create](https://man7.org/linux/man-pages/man2/memfd_create.2.html) system call and pass &[u8]s to safe Rust, since it's an anonymous open file which other processes wouldn't be able to access. However, without I/O safety, and without permanently sealing the file, other code in the program could accidentally call write or ftruncate on the file descriptor, breaking the memory-safety invariants of &[u8].

So safe code shouldn't be able to `write` or `truncate` an arbitrary fd, which could break immutability assumptions of something mmapping that fd. `dup` is also a problem, because safe code could write or truncate given an `OwnedFd`, which will have the same issue as write or truncate to the original fd.

Reply:
Sure, but then everything is unsafe because you can just open `/proc/self/mem`. I think we have to take a pragmatic approach here and making dup unsafe because someone might be messing around with mmap is overkill.

Reply:
Yeah, I think `dup2_raw()` should be unsafe, it is pretty easy to trigger double close with:


```rs
let file = File::open();
let owned_fd = dup2_raw(fd, file.as_raw_fd());
// double close
```